### ISSUE TITLE
Users can register using a taken email as username

### ISSUE THREAD

**Issue by WalkingPizza** [strapi/strapi#12301]
### Describe the bug

If user 1 registers an account on Strapi with email `foo@bar.com`, the user-permissions plugin will allow user 2 to register with username `foo@bar.com`. Since the login function accepts a generic `identifier` parameter instead of `username` or `email`, user 2 will not be able to log in via username.

### Steps to reproduce the behavior

1. Create user 1 with email `foo@bar.com`;
2. Create user 2 with username `foo@bar.com`;
3. Try logging in with `foo@bar.com` and user 2's password;
4. Notice how you cannot log in.

### Expected behavior

Return an error when users use a taken email as their username during registration.

### Screenshots

![image](https://user-images.githubusercontent.com/34224657/151071196-98d25d36-dffe-49fc-94d2-cdaf942bbe14.png)

### System

- Node.js version: 16.13.2
- NPM version: 8.1.2
- Yarn version: 1.22.15
- Strapi version: 4.0.5
- Database: PostgreSQL
- Operating system: Alpine

**Comment by petersg83** [comment]
What do you think of forbidding usernames to be an email ? It is simple to implement and would avoid any kind of confusion later on. To me it seems unlikely that people want to put an email for their username but maybe there are usecases I'm not aware about.

**Comment by WalkingPizza** [comment]
I would personally only allow alphanumeric characters, hyphens and dashes in usernames, and check that they don't start or end with a dash or hyphen. What do you think?

**Comment by petersg83** [comment]
Thank you for your insight! We are going to discuss about it.
On my personal opinion I think it is ok if it is up to the user to define what is possible for usernames. If some restrictions are needed, then the user can customize the code to make this validation beforehand.
So I would let it open.
I'm not even sure about forbidding email actually as it can be a breaking change. Maybe the way to go would be to just make sure the username is not the email of someone else by checking this condition during the registration phase and when users are modified.

**Comment by Convly** [comment]
> Maybe the way to go would be to just make sure the username is not the email of someone else by checking this condition during the registration phase and when users are modified.

That would leak the fact that this email is already taken by someone else though.
But yeah, this is already the case since you can already look for taken emails with the unique email validation so, I don't really know :shrug:

**Comment by derrickmehaffy** [comment]
> > Maybe the way to go would be to just make sure the username is not the email of someone else by checking this condition during the registration phase and when users are modified.
> 
> That would leak the fact that this email is already taken by someone else though. But yeah, this is already the case since you can already look for taken emails with the unique email validation so, I don't really know shrug

It's honestly not a security risk to see if an email is taken for example:

![image](https://user-images.githubusercontent.com/8593673/151461107-fb32855d-2ee9-4ddd-960c-5c16ad57a2bc.png)

Even google will happily tell you that a username/email is taken, I think the easiest solution is to just check:

- username doesn't have a duplicate with username or email
- email doesn't have a duplicate with username or email

We automatically lowercase everything (at least email I think) but we could force lowercase to check for duplicates and that should pretty much cover every case I can think of.

**Comment by petersg83** [comment]
Regarding this issue, we didn't find any solution that would not introduce other edge cases (currently username are automatically set when you use a provider, and for Microsoft the username is an email. So it doesn't make sense to forbid the use of email).
Since it is an edgecase, we think that we will let this bug as it is for now.

**Comment by WalkingPizza** [comment]
Wouldn't it make more sense to adjust the auth provider so that usernames are automatically set to whatever comes before the @ instead?

**Comment by petersg83** [comment]
Yes it would make more sense, it is a technical debt :/ changing this would be a breaking change that's why we prefer not to do that. At some points, because of the technical debts, we plan to replace this plugin with a better one

**Comment by petersg83** [comment]
Closing as we will not fix this issue
